# Bioinformatics Analysis of DNA Methylation

This repository provides a bioinformatics protocol for analyzing genome-wide DNA methylation data from **bisulfite sequencing (BS-seq)** and **enzymatic methyl sequencing (EM-seq)**.  

![DNA_methylation.png](https://github.com/PaoyangLab/Methylation_Analysis/blob/main/Figures/DNA_methylation.png)

The pipeline covers read alignment, methylation calling, DMR identification, visualization, and post-alignment analyses.  

---

## Table of Contents

- [1. Installation](#1-tools-installation)
- [2. Download Example Data](#2-download-example-data)
- [3. Analysis Pipeline](#3-analysis-pipeline)
	- [3.1 Processing methylomes](#31-processing-methylomes)
	- [3.2 DMR identification](#32-dmr-identification)
	- [3.3 Data visualization](#33-data-visualization)
	- [3.4 Post-alignment analyses](#34-post-alignment-analyses)

---

## 1. Tools Installation

### From GitHub
- [SRA Toolkit 3.0.5](https://github.com/ncbi/sra-tools)  
- [BS-Seeker2 v2.0.8](https://github.com/BSSeeker/BSseeker2)  
- [HOME v1.0.0](https://github.com/ListerLab/HOME)  
- [MethylC-analyzer](https://github.com/RitataLU/MethylC-analyzer)  

> For the GitHub tools, please download the repository through `clone` command and install them by following the instructions in their manuals.
```bash
git clone <GIT_URL> 
```

> For MethylC-analyzer, Docker image is recommended to avoid environment conflict
```bash
docker pull peiyulin/methylc:V1.0
```

###  From Other Sites 
- [Bowtie2 v2.26](https://bowtie-bio.sourceforge.net/bowtie2/index.shtml)  
- [bicycle v1.8.2](http://www.sing-group.org/bicycle)  
- [IGV Desktop v2.16.0](https://igv.org/)  
- [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/)  

> For these tools, you can download them through `wget` command and install them by following the instructions in their manuals.
```bash
wget <TOOL_URL> 
```

---

## 2. Download Example Data

We demonstrate the pipeline using ***Arabidopsis thaliana*** dataset [GSE122394](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE122394) from GEO, including wild-type (wt) strains as controls and *met1* mutant strains in which DNA methyltransferase 1 (MET1) functions primarily to maintain CG methylation. Each group (wild-type, met1 mutant) contains **three biological replicates**.  

|Name|SRA|Type|
|----|---|----|
|wt_r1|SRR8180314|Wild type|
|wt_r2|SRR8180315|Wild type|
|wt_r3|SRR8180316|Wild type|
|met1_r1|SRR8180322|MET1 mutant|
|met1_r2|SRR8180323|MET1 mutant|
|met1_r3|SRR8180324|MET1 mutant|

To obtain the data, you can use SRA Toolkit to download the file by `prefetch` and then convert it into the FASTQ format (.fastq) for analysis by `fast-dump`.

```bash
prefetch <SRA_ID> ## download SRA data
fastq-dump <SRA_ID> ## transfer into fastq file 
```

Reference genome (TAIR10) can be downloaded from [iGenomes](https://support.illumina.com/sequencing/sequencing_software/igenome.html).

---

## 3. Analysis Pipeline

The bioinformatics pipeline of methylation analysis is introduced below, including Processing methylomes, DMR identification, Data visualization and Post-alignment analyses.

![overall_pipeline.png](https://github.com/PaoyangLab/Methylation_Analysis/blob/main/Figures/overall_pipeline.png)

### 3.1 Processing methylomes
Here, BS-Seeker2 is used to align reads and call methylation. Replicate 1 of the wild-type sample is used as an example.

#### 3.1.1 Quality control and Remove duplicates
Before alignment, the methyl-seq reads should undergo quality control (QC) to remove low-quality reads and duplicate sequences generated by PCR amplification and adapter sequences. The suggested tool for QC is [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/), and BS-Seeker2 also provides the command lines for removing duplicated reads.

1. Quality control
> Fastqc will generate QC report to check read quality.
```bash
fastqc wt_r1.fastq
```
2. Remove duplicate
> The script `FilterReads.py` can be found from [BS-Seeker2](https://github.com/BSSeeker/BSseeker2/blob/master/FilterReads.py). The `-i` specifies the input FASTQ file of WT replicate 1 `wt_r1.fastq`, which is obtained from [Step 2](#2-download-example-data); `-o` specifies the output file `wt_r1_rmdup.fastq`; `FilterReads.log` records the processing log.
```bash
FilterReads.py -i wt_r1.fastq -o wt_r1_rmdup.fastq > FilterReads.log
```

#### 3.1.2 Alignment of methyl-seq reads
1. Use bowtie2 to create a reference genome index file (Arabidopsis thaliana TAIR10 version) for the aligner. 
> 	The `-f` specifies the FASTA file of reference genome `genome.fa`, which can be downloaded from [iGenomes](https://support.illumina.com/sequencing/sequencing_software/igenome.html). The `-d` specifies the directory to save output index file.
```bash
bs_seeker2-build.py -f genome.fa --aligner=bowtie2 -d ./BS2_bt2_Index
```

2. Align raw reads of wild-type replicate 1 to the reference genome.
> 	The `-i` specifies input FASTQ file of WT replicate 1 `wt_r1_rmdup.fastq`, which is obtained from [Step 3.1.1](#311-quality-control-and-remove-duplicates); the `-g` specifies reference genome `genome.fa` obtained from previous step; and `-o` specifies output BAM file named `wt_r1_align.bam`.
```bash
bs_seeker2-align.py -i wt_r1_rmdup.fastq -g genome.fa  --aligner=bowtie2 -o wt_r1_align.bam
```

#### 3.1.3 Call methylation
1. Use methylation-calling function to calculate the methylation level. 
> 	Input BAM file `wt_r1_align.bam` is obtained from [Step 3.1.2](#312-alignments-of-methyl-seq-reads). Output file is saved as a CGmap file named `wt_r1` (the output file will be zipped into `wt_r1.CGmap.gz`). The `-d` parameter is used to specifies the index file of reference genome.
```bash
bs_seeker2-call_methylation.py -i wt_r1_align.bam -o wt_r1 -d ./BS2_bt2_Index/genome.fa_bowtie2
```

2. View the methylation call output (CGmap). The file with each row represents a single CpG site.
```bash
zless wt_r1.CGmap.gz
```

Each CpG site contains the following information: chromosome, nucleotide on Watson strand, position, context, dinucleotide context, methylation level, number of methylated cytosines ($N_C$), and the total number of all cytosines ($N_T$ + $N_C$)

![CpGmap_table.png](https://github.com/PaoyangLab/Methylation_Analysis/blob/main/Figures/CpGmap_table.png)

#### 3.1.4 Conversion rate
In regard to methyl-seq (EM-seq and BS-seq) analysis, the estimation conversion rate, which measures how effectively bisulfite or enzyme treatment can convert unmethylated cytosines to uracil in DNA samples, is required for evaluation. By comparing the unmethylated bacteriophage lambda genome as a reference to our bisulfite/enzyme treatment genomes, the percentage of successfully converted cytosines can be estimated. It is simply calculated by dividing the number of **converted cytosines ($N_T$)** by the **total number of cytosines ($N_T$ + $N_C$)** and multiplying by 100. 

$$
\text{Conversion rate} = \frac{N_T}{N_T + N_C} \times 100\%
$$



Typically, a conversion rate of 95% or above is preferred because it shows more reliable and accurate results.

1. The first step for the conversion rate is the same as above but changes the input reference genome to the lambda phage genome `lambda_genome.fa`.
```bash
bs_seeker2-build.py -f lambda_genome.fa --aligner=bowtie2 -d ./BS2_lambda_Index

bs_seeker2-align.py -i wt_r1_rmdup.fastq -g lambda_genome.fa  --aligner=bowtie2 -o wt_r1_lambda.bam -m 3 -d BS2_lambda_Index

bs_seeker2-call_methylation.py -i wt_r1_lambda.bam -o wt_r1_lambda -d BS2_lambda_Index/lambda_genome.fa_bowtie2/
```

2. The conversion rate is calculated by the R script with the formula. 
> 	The script [conversion_rate.R ](https://github.com/PaoyangLab/Methylation_Analysis/blob/main/conversion_rate.R) is included in this repository.
```bash
Rscript conversion_rate.R  wt_r1_lambda.CGmap.gz
```
> The output will look like:
```bash
[ 03:13:46 AM ] Calculating bisulfite conversion rate
[ 03:13:46 AM ] Bisulfite conversion rate: 97.01493 %
```

In our example, the conversion rate for the wt_r1 methylome is 97.01%, which means that 97.01% of the unmethylated cytosines in the DNA sample have been successfully converted to uracil.

### 3.2 DMR identification
Here, MethylC-analyzer is selected to demonstrate how to find DMRs from the aligned methylation data output. To prevent environmental conflicts, the Docker image provided by the software is utilized
#### 3.2.1 Searching DMR
1. Prepare input files 
	- List CGmaps into TXT file `samples_list.txt` (obtained from [Step 3.1.3](#313-call-methylation))
		```bash
		wt_r1.CGmap.gz
		wt_r2.CGmap.gz
		wt_r3.CGmap.gz
		met1_r1.CGmap.gz
		met1_r2.CGmap.gz
		met1_r3.CGmap.gz
		```
	- Gene annotation GTF files `gene.gtf` can be downloaded from UCSC (https://hgdownload.soe.ucsc.edu/downloads.html), and it is a file format containing information about the genomic features of genes, such as exons, introns, coding sequences, and untranslated regions (UTRs).

2. Run MethylC-analyzer to identify the DMRs
> 	The default minimum depth for CpG sites and the number of sites within a region are both set to **4**. The default size of the DMR is **500 base pairs (bp)**. The default p-value cutoff for Studentâ€™s t-test for identifying DMRs is **P<0.05**. These arguments can be adjusted by users. The `-a` and `-b` specifies the group names.
```bash
docker run --rm -v $(pwd):/app peiyulin/methylc:V1.0 python /MethylC-analyzer/scripts/MethylC.py DMR samples_list.txt gene.gtf /app/ -a met1 -b wt
```

The output consists of all, hyper, and hypo DMRs as text files. Here, we found **3,282 DMRs** in CG methylation between the wt and met1 groups. 

### 3.3 Data visualization
#### 3.3.1 Genome browser

![IGV_tutorial 1.png](https://github.com/PaoyangLab/Methylation_Analysis/blob/main/Figures/IGV_tutorial.png)

1.	Download and activate the IGV Desktop application according to the operating system. This application supports operating systems including macOS, Windows, and Linux.
2.	Select the reference genome from the dropdown list. Here, we chose A. thaliana (TAIR10) as a reference genome. Additional reference genomes can be downloaded by clicking More or can be loaded from the local path (in FASTA format).
3.	Convert the file from the WIG file to the suggested track formats, BigWig or TDF files, by running IGVtools (Click `Tools`>`Run IGVtools`).
4.	Select `File`>`Load from File` to load data into the track panel. Right-click the panel to adjust the graphic type or other settings.
5.	Use the dropdown list and search box at the top panel to select the chromosome and region shown. Click `+`/`-` on the top panel to zoom in/out. Clicking or dragging on the track of the chromosome can also adjust the region shown.
6.	Click `File`>`Save session or File`>`Save Image` to save the visualization result.

### 3.4 Post-alignment analyses
MethylC-analyzer provides several post-alignment analyses. Here, we implement the enrichment analysis and metagene analysis as examples.

![methylC_tutorial.png](https://github.com/PaoyangLab/Methylation_Analysis/blob/main/Figures/methylC_tutorial.png)
#### 3.4.1 Enrichment analysis
Use the `Fold_Enrichment` command to generate the enrichment result. 
This module generates output files, including `CG_Fold_Enrichment.pdf` and multiple BED files, such as `CommonRegion_CG.txt.bed`. The BED format provides the information like the positions of common methylated regions across samples. The BED file can be visualized by using IGV.

```bash
docker run --rm -v $(pwd):/app peiyulin/methylc:V1.0 python /MethylC-analyzer/scripts/MethylC.py Fold_Enrichment samples_list.txt gene.gtf /app/ -a met1 -b wt
```

DMRs exhibit a positive fold enrichment value in the IGR, suggesting a higher likelihood of DMRs being located in IGRs.

<img src="https://github.com/PaoyangLab/Methylation_Analysis/blob/main/Figures/CG_Fold_Enrichment.png" width="400">

#### 3.4.2 Metagene analysis
 Use the `Metaplot` command to generate the Metaplot result. This module generates two types of metagene plots: one represents the average methylation level in two groups (metaplot_CG.pdf), and the other shows the difference 
between the two groups (metaplot_delta_CG.pdf). The former illustrates the methylation pattern along the gene body and adjacent region, while the latter directly represents the difference in distribution between wt and met1. This module also generates BigWig files (met1_r1_CG.bw) to record methylated C sites in metagene analysis, and these BigWig files can be visualized by IGV. 

```bash
docker run --rm -v $(pwd):/app peiyulin/methylc:V1.0 python /MethylC-analyzer/scripts/MethylC.py Metaplot samples_list.txt gene.gtf /app/ -a met1 -b wt
```

In our case, the wt samples exhibit a standard CG methylation pattern with a lower methylation level at the transcription start site (TSS) and transcription end site (TES). The met1 samples show a consistently low methylation level along the gene body, reflecting the dysfunction of the methyltransferase.

<img src="https://github.com/PaoyangLab/Methylation_Analysis/blob/main/Figures/metaplot_CG.png" width="400">

